core/services/workflows/metering/balance_store.go

```
@@ -48,123 +53,147 @@ func NewBalanceStore(startingBalance int64, conversions map[string]decimal.Decim
+		conversions:  conversions,
+		lggr:         lggr,
 	}
 }
 
 // convertToBalance converts a resource dimension amount to a credit amount.
 // This method should only be used under a read lock.
-func (bs *balanceStore) convertToBalance(fromUnit string, amount int64) (credits int64) {
+func (bs *balanceStore) convertToBalance(fromUnit string, amount decimal.Decimal) decimal.Decimal {
 	rate, ok := bs.conversions[fromUnit]
 	if !ok {
 		// Fail open, continue optimistically
-		bs.lggr.Errorw("could not find conversion rate, continuing as 1:1", "unit", fromUnit)
+		bs.lggr.Errorw("could not find conversion rate, continuing as 1:1; entering metering mode", "unit", fromUnit)
 		rate = decimal.NewFromInt(1)
+		bs.meteringMode = true
 	}
-	return decimal.NewFromInt(amount).Mul(rate).RoundUp(0).IntPart()
+
+	return amount.Mul(rate)
 }
 
 // ConvertToBalance converts a resource dimensions amount to a credit amount.
-func (bs *balanceStore) ConvertToBalance(fromUnit string, amount int64) (credits int64) {
+func (bs *balanceStore) ConvertToBalance(fromUnit string, amount decimal.Decimal) decimal.Decimal {
 	bs.mu.RLock()
 	defer bs.mu.RUnlock()
+
 	return bs.convertToBalance(fromUnit, amount)
 }
 
 // convertFromBalance converts a credit amount to a resource dimensions amount.
 // This method should only be used under a read lock.
-func (bs *balanceStore) convertFromBalance(toUnit string, amount int64) (resources int64) {
+func (bs *balanceStore) convertFromBalance(toUnit string, amount decimal.Decimal) decimal.Decimal {
 	rate, ok := bs.conversions[toUnit]
 	if !ok {
 		// Fail open, continue optimistically
-		bs.lggr.Errorw("could not find conversion rate, continuing as 1:1", "unit", toUnit)
+		bs.lggr.Errorw("could not find conversion rate, continuing as 1:1; entering metering mode", "unit", toUnit)
 		rate = decimal.NewFromInt(1)
+		bs.meteringMode = true
 	}
-	return decimal.NewFromInt(amount).Div(rate).RoundUp(0).IntPart()
+
+	return amount.Div(rate)
```

if we enter meteringMode, is it meaningful to have a return value? I'd argue no - the way metering mode stands today, the balanceStore no longer contributes to the runtime flow of the execution. We should just be collecting the resource spends and populating the metering report. 

Concretely, what about changing the API to:

```
func (bs *balanceStore) convertFromBalance(toUnit string, amount decimal.Decimal) (resourceAmount decimal.Decimal, err error) {
	rate, ok := bs.conversions[toUnit]
	if !ok {
		// Fail open, continue optimistically
		bs.lggr.Errorw("could not find conversion rate; entering metering mode", "unit", toUnit)
		rate = decimal.NewFromInt(1)
		bs.meteringMode = true
	}

	return amount.Div(rate)
```

and then 

```
func (bs *balanceStore) ConvertToBalance(fromUnit string, amount decimal.Decimal) (resourceAmount decimal.Decimal, err error) {
	bs.mu.RLock()
	defer bs.mu.RUnlock()

	return bs.convertToBalance(fromUnit, amount)
}
```

---

What about even pulling `meteringMode` up to the struct of `meteringReport`. Then ignore errors on that level. That would leave `balanceStore` more explicit to when we _are_ tracking balance.

---

It's not meaningful to have a return value, but I think likewise it's not meaningful to have an error either. The error is essentially handled and a value is returned whether it has meaning or not.